File: TimePilot-IIvera/tools/build_hdv.mjs
// build_hdv.mjs — Package Time Pilot IIvera into a bootable ProDOS 8 HDV (.hdv).
//
// Boot chain: ProDOS HDV -> BASIC.SYSTEM -> Applesoft STARTUP -> BRUN MAIN.BIN.
// Audio (pcm.blob) is placed at FIXED blocks (PCM_START_BLOCK) and read by the
// 6502 directly via MLI READ_BLOCK — it is NOT a named ProDOS file (the veratest
// "fixed music slot" pattern). Only MAIN.BIN and STARTUP get directory entries.
//
// Modeled on C:\dev\veratest\src\slideshow\slideshow_hdv.mjs (proven on apple2ts).
import fs from "fs"
import path from "path"
import { fileURLToPath } from "url"
import { compileApplesoftBasic } from "../../../veratest/src/applebasic.mjs"

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const projectRoot = path.resolve(__dirname, "..")
const srcDir = path.join(projectRoot, "src")
const buildDir = path.join(projectRoot, "build")
const baseHdvPath = "C:/dev/veratest/assets/ProDOS 2.4.3.hdv"
const BLOCK = 512

if (!fs.existsSync(baseHdvPath)) throw new Error(`Base HDV not found: ${baseHdvPath}`)
if (!fs.existsSync(path.join(buildDir, "main.bin"))) throw new Error("build/main.bin missing — run build.bat first")
if (!fs.existsSync(path.join(buildDir, "pcm.blob"))) throw new Error("build/pcm.blob missing — run tools/mkpcm_blob.mjs first")

const disk = new Uint8Array(fs.readFileSync(baseHdvPath))
const TOTAL = disk.length / BLOCK
const mainBinRaw = new Uint8Array(fs.readFileSync(path.join(buildDir, "main.bin")))
// Strip the 4-byte ProDOS BIN header (load addr LE + length LE). BRUN loads at
// the aux-type address and the file content must be raw code starting at byte 0.
// (Header: 00 20 = $2000 LE, then length LE. Code follows.)
const mainBin = mainBinRaw.subarray(4)
const pcm = new Uint8Array(fs.readFileSync(path.join(buildDir, "pcm.blob")))

// ---- Audio: fixed block range (keep in sync with mkpcm_blob.mjs PCM_START_BLOCK) ----
const PCM_START_BLOCK = 200
const PCM_BLOCKS = Math.ceil(pcm.length / BLOCK)
{
  const start = PCM_START_BLOCK * BLOCK
  disk.set(pcm, start)
  // pad the final block with zeros (MLI reads whole blocks)
  disk.fill(0, start + pcm.length, (PCM_START_BLOCK + PCM_BLOCKS) * BLOCK)
}
console.log(`  audio: pcm.blob ${pcm.length}B at blocks ${PCM_START_BLOCK}..${PCM_START_BLOCK + PCM_BLOCKS - 1}`)

// ---- Art: fixed block range (keep in sync with mkart.mjs ART_START_BLOCK) ----
const artPath = path.join(buildDir, "art.blob")
let ART_START_BLOCK = 0, ART_BLOCKS = 0
if (fs.existsSync(artPath)) {
  const art = new Uint8Array(fs.readFileSync(artPath))
  ART_START_BLOCK = 900
  ART_BLOCKS = Math.ceil(art.length / BLOCK)
  const start = ART_START_BLOCK * BLOCK
  disk.set(art, start)
  disk.fill(0, start + art.length, (ART_START_BLOCK + ART_BLOCKS) * BLOCK)
  console.log(`  art:   art.blob ${art.length}B at blocks ${ART_START_BLOCK}..${ART_START_BLOCK + ART_BLOCKS - 1}`)
}

// ---- Demo: fixed block range (keep in sync with demo_data.h DEMO_START_BLOCK) ----
const demoPath = path.join(buildDir, "demo.blob")
let DEMO_START_BLOCK = 0, DEMO_BLOCKS = 0
if (fs.existsSync(demoPath)) {
  const demo = new Uint8Array(fs.readFileSync(demoPath))
  DEMO_START_BLOCK = 980
  DEMO_BLOCKS = Math.ceil(demo.length / BLOCK)
  const start = DEMO_START_BLOCK * BLOCK
  disk.set(demo, start)
  disk.fill(0, start + demo.length, (DEMO_START_BLOCK + DEMO_BLOCKS) * BLOCK)
  console.log(`  demo:  demo.blob ${demo.length}B at blocks ${DEMO_START_BLOCK}..${DEMO_START_BLOCK + DEMO_BLOCKS - 1}`)
}

// ---- Block allocator (skip 0..99 system reserve + audio + art + demo ranges) ----
const used = new Set()
for (let b = 0; b < 100; b++) used.add(b)
for (let b = PCM_START_BLOCK; b < PCM_START_BLOCK + PCM_BLOCKS; b++) used.add(b)
for (let b = ART_START_BLOCK; b < ART_START_BLOCK + ART_BLOCKS; b++) used.add(b)
if (DEMO_BLOCKS > 0) {
  for (let b = DEMO_START_BLOCK; b < DEMO_START_BLOCK + DEMO_BLOCKS; b++) used.add(b)
}
const newlyAllocated = []   // for the bitmap update
let nextFree = 100
const allocate = () => {
  while (used.has(nextFree)) nextFree++
  const b = nextFree++
  used.add(b)
  newlyAllocated.push(b)
  return b
}

// Write a ProDOS file (type 1 = single block, type 2 = index + data blocks).
function writeFile(name, fileType, aux, data) {
  const size = data.length
  let stType, keyBlock, totalBlocks
  if (size <= BLOCK) {
    stType = 1
    keyBlock = allocate()
    disk.set(data, keyBlock * BLOCK)
    totalBlocks = 1
  } else {
    stType = 2
    keyBlock = allocate()
    const idx = new Uint8Array(BLOCK)
    const n = Math.ceil(size / BLOCK)
    for (let i = 0; i < n; i++) {
      const db = allocate()
      disk.set(data.subarray(i * BLOCK, Math.min(size, (i + 1) * BLOCK)), db * BLOCK)
      idx[i] = db & 0xFF
      idx[i + 256] = (db >> 8) & 0xFF
    }
    disk.set(idx, keyBlock * BLOCK)
    totalBlocks = 1 + n
  }
  return { name, stType, fileType, keyBlock, totalBlocks, eof: size, aux }
}

// Register a range of ALREADY-PLACED data blocks (at fixed addresses the game
// reads directly via MLI) as one or more ProDOS "sapling" files so CATALOG can
// list them — same pattern as veratest's /DATA images. A single ProDOS file has
// at most 256 index slots (=128KB), so larger blobs are split into parts. The
// data blocks themselves are NOT moved; we only add an index block + dir entry.
function addDataFile(name, fileType, baseBlock, numBlocks) {
  const entries = []
  let offset = 0, part = 1
  while (offset < numBlocks) {
    const chunk = Math.min(256, numBlocks - offset)
    const idxBlock = allocate()
    const idx = new Uint8Array(BLOCK)
    for (let i = 0; i < chunk; i++) {
      const db = baseBlock + offset + i
      idx[i] = db & 0xFF
      idx[i + 256] = (db >> 8) & 0xFF
    }
    disk.set(idx, idxBlock * BLOCK)
    entries.push({
      name: numBlocks > 256 ? `${name}${String(part).padStart(1, "0")}` : name,
      stType: 2, fileType, keyBlock: idxBlock,
      totalBlocks: 1 + chunk, eof: chunk * BLOCK, aux: 0x2000
    })
    offset += chunk
    part++
  }
  return entries
}

const startupBas = path.join(srcDir, "startup.bas")
if (!fs.existsSync(startupBas)) throw new Error(`startup.bas not found at ${startupBas}`)
const startupBytes = new Uint8Array(compileApplesoftBasic(srcDir, "startup.bas"))
const fStartup = writeFile("STARTUP", 0xFC, 0x0801, startupBytes)
const mainLoadAddr = mainBinRaw[0] | (mainBinRaw[1] << 8)
const fMain = writeFile("MAIN.BIN", 0x06, mainLoadAddr, mainBin)
console.log(`  MAIN.BIN  ${mainBin.length}B (load=$${mainLoadAddr.toString(16).toUpperCase()}, key=${fMain.keyBlock}, ${fMain.totalBlocks} blocks)`)
const appFiles = [fStartup, fMain]
const main4Path = path.join(buildDir, "main4.bin")
if (fs.existsSync(main4Path)) {
  const raw4 = new Uint8Array(fs.readFileSync(main4Path))
  const main4LoadAddr = raw4[0] | (raw4[1] << 8)
  const main4Bin = raw4.subarray(4)
  const fMain4 = writeFile("MAIN4.BIN", 0x06, main4LoadAddr, main4Bin)
  appFiles.push(fMain4)
  console.log(`  MAIN4.BIN ${main4Bin.length}B (load=$${main4LoadAddr.toString(16).toUpperCase()}, key=${fMain4.keyBlock}, ${fMain4.totalBlocks} blocks)`)
}
console.log(`  STARTUP   ${startupBytes.length}B (key=${fStartup.keyBlock}, ${fStartup.totalBlocks} blocks)`)

// ---- Register the audio + art blobs as ProDOS files so CATALOG lists them.
// Data stays at its fixed block range; we only add index blocks + dir entries.
const dataFiles = []
dataFiles.push(...addDataFile("PCM", 0x06, PCM_START_BLOCK, PCM_BLOCKS))
if (ART_BLOCKS > 0) dataFiles.push(...addDataFile("ART", 0x06, ART_START_BLOCK, ART_BLOCKS))
if (DEMO_BLOCKS > 0) dataFiles.push(...addDataFile("DEMO", 0x06, DEMO_START_BLOCK, DEMO_BLOCKS))
for (const e of dataFiles) console.log(`  ${e.name.padEnd(8)} ${e.eof}B (${e.totalBlocks} blocks, key=${e.keyBlock})`)

// ---- Rewrite the root directory (block 2) ----
const vol = disk.subarray(2 * BLOCK, 3 * BLOCK)
const nameOf = (off, len) => String.fromCharCode(...vol.subarray(off + 1, off + 1 + len))
// Preserve PRODOS / CLOCK.SYSTEM / BASIC.SYSTEM; drop the rest (BITSY.BOOT, QUIT.SYSTEM).
const keep = []
for (let i = 1; i <= 12; i++) {
  const off = 4 + i * 39
  if (vol[off] === 0) continue
  const nameLen = vol[off] & 0x0F
  const name = nameOf(off, nameLen)
  if (name === "PRODOS" || name === "CLOCK.SYSTEM" || name === "BASIC.SYSTEM") keep.push(vol.slice(off, off + 39))
}
// Clear all file entries.
for (let i = 1; i <= 12; i++) vol.fill(0, 4 + i * 39, 4 + (i + 1) * 39)

// Write back the preserved system files verbatim, then the new app files.
let idx = 1
for (const raw of keep) {
  vol.set(raw, 4 + idx * 39)
  idx++
}
for (const e of [...appFiles, ...dataFiles]) {
  const off = 4 + idx * 39
  vol[off] = (e.stType << 4) | (e.name.length & 0x0F)
  vol.set([...e.name].map((c) => c.charCodeAt(0)), off + 1)
  vol[off + 0x10] = e.fileType
  vol[off + 0x11] = e.keyBlock & 0xFF
  vol[off + 0x12] = (e.keyBlock >> 8) & 0xFF
  vol[off + 0x13] = e.totalBlocks & 0xFF
  vol[off + 0x14] = (e.totalBlocks >> 8) & 0xFF
  vol[off + 0x15] = e.eof & 0xFF
  vol[off + 0x16] = (e.eof >> 8) & 0xFF
  vol[off + 0x17] = (e.eof >> 16) & 0xFF
  vol[off + 0x1E] = 0xC3
  vol[off + 0x1F] = e.aux & 0xFF
  vol[off + 0x20] = (e.aux >> 8) & 0xFF
  vol[off + 0x25] = 2          // parent = volume block 2
  vol[off + 0x26] = 0
  idx++
}
const entryCount = keep.length + appFiles.length + dataFiles.length
// Volume file count (entries after the header).
vol[0x25] = entryCount & 0xFF
vol[0x26] = (entryCount >> 8) & 0xFF

// ---- Mark newly-allocated + audio blocks as used in the ProDOS bitmap ----
// (Blocks 6..21 hold the 16-bit bitmap; bit set = FREE, clear = USED.)
const setUsed = (b) => {
  const bmpBlock = 6 + Math.floor(b / 64)
  const byteIdx = Math.floor((b % 64) / 8)
  const bit = 7 - ((b % 64) % 8)
  disk[bmpBlock * BLOCK + byteIdx] &= ~(1 << bit)
}
for (const b of newlyAllocated) setUsed(b)
for (let b = PCM_START_BLOCK; b < PCM_START_BLOCK + PCM_BLOCKS; b++) setUsed(b)
if (ART_BLOCKS > 0) for (let b = ART_START_BLOCK; b < ART_START_BLOCK + ART_BLOCKS; b++) setUsed(b)

const outPath = path.join(projectRoot, "TimePilot-IIvera.hdv")
fs.writeFileSync(outPath, disk)
console.log(`\n  Built ${outPath} (${disk.length} bytes = ${(disk.length / 1024 / 1024).toFixed(1)} MB)`)
console.log(`  Root files: ${keep.length} system + ${appFiles.map(f => f.name).join(" + ")}`)

File: TimePilot-IIvera/tools/build_hdv.py
#!/usr/bin/env python3
"""build_hdv.py - Package Time Pilot IIvera into a bootable ProDOS 8 HDV.

Boot chain: ProDOS -> CLOCK.SYSTEM -> TPILOT.SYSTEM -> MAIN.BIN at $0800.

Audio (pcm.blob) is placed at FIXED blocks (PCM_START_BLOCK) and read by the
6502 directly via MLI READ_BLOCK -- it is not a named ProDOS file. Only
TPILOT.SYSTEM, MAIN.BIN and MAIN4.BIN get real directory entries; PCM and ART
get index blocks pointing at their fixed data so CATALOG lists them.
"""

import argparse
import os
import sys

BLOCK = 512
PCM_START_BLOCK = 200   # keep in sync with offline/mkpcm_blob.mjs
ART_START_BLOCK = 900   # keep in sync with mkart.py
SYSTEM_RESERVE = 100    # blocks 0..99 belong to ProDOS + system files
KEEP_FILES = ("PRODOS", "CLOCK.SYSTEM")


class Allocator:
    """ProDOS block allocator that skips the system reserve and fixed ranges."""

    def __init__(self, reserved):
        self.used = set(reserved)
        self.newly_allocated = []
        self.next_free = SYSTEM_RESERVE

    def allocate(self):
        while self.next_free in self.used:
            self.next_free += 1
        b = self.next_free
        self.next_free += 1
        self.used.add(b)
        self.newly_allocated.append(b)
        return b


def write_file(disk, alloc, name, file_type, aux, data):
    """Write a ProDOS file: seedling (1 block) or sapling (index + data)."""
    size = len(data)
    if size <= BLOCK:
        st_type = 1
        key_block = alloc.allocate()
        disk[key_block * BLOCK:key_block * BLOCK + size] = data
        total_blocks = 1
    else:
        st_type = 2
        key_block = alloc.allocate()
        idx = bytearray(BLOCK)
        n = (size + BLOCK - 1) // BLOCK
        for i in range(n):
            db = alloc.allocate()
            chunk = data[i * BLOCK:min(size, (i + 1) * BLOCK)]
            disk[db * BLOCK:db * BLOCK + len(chunk)] = chunk
            idx[i] = db & 0xFF
            idx[i + 256] = (db >> 8) & 0xFF
        disk[key_block * BLOCK:(key_block + 1) * BLOCK] = idx
        total_blocks = 1 + n
    return dict(name=name, st_type=st_type, file_type=file_type,
                key_block=key_block, total_blocks=total_blocks,
                eof=size, aux=aux)


def add_data_file(disk, alloc, name, file_type, base_block, num_blocks):
    """Register already-placed blocks as ProDOS files without moving them.

    A single ProDOS index block holds at most 256 entries (128KB), so larger
    blobs are split into numbered parts.
    """
    entries = []
    offset = 0
    part = 1
    while offset < num_blocks:
        chunk = min(256, num_blocks - offset)
        idx_block = alloc.allocate()
        idx = bytearray(BLOCK)
        for i in range(chunk):
            db = base_block + offset + i
            idx[i] = db & 0xFF
            idx[i + 256] = (db >> 8) & 0xFF
        disk[idx_block * BLOCK:(idx_block + 1) * BLOCK] = idx
        entries.append(dict(
            name=(f"{name}{part}" if num_blocks > 256 else name),
            st_type=2, file_type=file_type, key_block=idx_block,
            total_blocks=1 + chunk, eof=chunk * BLOCK, aux=0x2000))
        offset += chunk
        part += 1
    return entries


def place_blob(disk, data, start_block):
    """Copy a blob to a fixed block range and zero-pad its final block."""
    n = (len(data) + BLOCK - 1) // BLOCK
    start = start_block * BLOCK
    disk[start:start + len(data)] = data
    for i in range(start + len(data), (start_block + n) * BLOCK):
        disk[i] = 0
    return n


def write_dir_entry(vol, off, e):
    vol[off] = (e["st_type"] << 4) | (len(e["name"]) & 0x0F)
    for i, ch in enumerate(e["name"]):
        vol[off + 1 + i] = ord(ch)
    vol[off + 0x10] = e["file_type"]
    vol[off + 0x11] = e["key_block"] & 0xFF
    vol[off + 0x12] = (e["key_block"] >> 8) & 0xFF
    vol[off + 0x13] = e["total_blocks"] & 0xFF
    vol[off + 0x14] = (e["total_blocks"] >> 8) & 0xFF
    vol[off + 0x15] = e["eof"] & 0xFF
    vol[off + 0x16] = (e["eof"] >> 8) & 0xFF
    vol[off + 0x17] = (e["eof"] >> 16) & 0xFF
    vol[off + 0x1E] = 0xC3
    vol[off + 0x1F] = e["aux"] & 0xFF
    vol[off + 0x20] = (e["aux"] >> 8) & 0xFF
    vol[off + 0x25] = 2          # parent = volume directory block 2
    vol[off + 0x26] = 0


def main():
    here = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(here)

    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--src-dir", default=os.path.join(project_root, "src"))
    ap.add_argument("--assets-dir", default=os.path.join(project_root, "assets"))
    ap.add_argument("--build-dir", default=os.path.join(project_root, "build"))
    ap.add_argument("--out-dir", default=os.path.join(project_root, "disks"))
    ap.add_argument("--out-name", default="TimePilot-IIvera.hdv")
    args = ap.parse_args()

    base_hdv = os.path.join(args.assets_dir, "800kb.hdv")
    pcm_path = os.path.join(args.assets_dir, "pcm.blob")
    art_path = os.path.join(args.build_dir, "art.blob")
    sys_path = os.path.join(args.build_dir, "tpilot.sys")
    main_path = os.path.join(args.build_dir, "main.bin")
    main4_path = os.path.join(args.build_dir, "main4.bin")

    for p, what in ((base_hdv, "base HDV template"),
                    (pcm_path, "PCM audio asset"),
                    (sys_path, "compiled tpilot.sys"),
                    (main_path, "compiled main.bin")):
        if not os.path.exists(p):
            raise SystemExit(f"error: {what} not found: {p}")

    with open(base_hdv, "rb") as f:
        disk = bytearray(f.read())
    with open(pcm_path, "rb") as f:
        pcm = f.read()
    with open(main_path, "rb") as f:
        main_raw = f.read()

    # Strip the 4-byte ProDOS BIN header (load addr LE + length LE).
    # TPILOT.SYSTEM streams the raw payload to $0800; aux_type records that.
    main_bin = main_raw[4:]
    main_load_addr = main_raw[0] | (main_raw[1] << 8)

    pcm_blocks = place_blob(disk, pcm, PCM_START_BLOCK)
    print(f"  audio: pcm.blob {len(pcm)}B at blocks "
          f"{PCM_START_BLOCK}..{PCM_START_BLOCK + pcm_blocks - 1}")

    art_blocks = 0
    if os.path.exists(art_path):
        with open(art_path, "rb") as f:
            art = f.read()
        art_blocks = place_blob(disk, art, ART_START_BLOCK)
        print(f"  art:   art.blob {len(art)}B at blocks "
              f"{ART_START_BLOCK}..{ART_START_BLOCK + art_blocks - 1}")

    reserved = set(range(SYSTEM_RESERVE))
    reserved |= set(range(PCM_START_BLOCK, PCM_START_BLOCK + pcm_blocks))
    reserved |= set(range(ART_START_BLOCK, ART_START_BLOCK + art_blocks))
    alloc = Allocator(reserved)

    with open(sys_path, "rb") as f:
        sys_raw = f.read()
    # llvm-mos may write a sibling .elf; the -o path is the raw SYS image.
    if sys_raw[:4] == b"\x7fELF":
        raise SystemExit("error: tpilot.sys looks like an ELF, not a ProDOS SYS image")
    # mos-apple2-clang emits a 4-byte ProDOS BIN header ($2000 + length).
    # ProDOS SYS (file type $FF) files are loaded directly at $2000 without a header.
    if len(sys_raw) >= 4 and (sys_raw[0] | (sys_raw[1] << 8)) == 0x2000:
        sys_raw = sys_raw[4:]

    f_sys = write_file(disk, alloc, "TPILOT.SYSTEM", 0xFF, 0x2000, sys_raw)
    f_main = write_file(disk, alloc, "MAIN.BIN", 0x06, main_load_addr, main_bin)
    print(f"  TPILOT.SYSTEM {len(sys_raw)}B (key={f_sys['key_block']}, "
          f"{f_sys['total_blocks']} blocks)")
    print(f"  MAIN.BIN  {len(main_bin)}B (load=${main_load_addr:X}, "
          f"key={f_main['key_block']}, {f_main['total_blocks']} blocks)")

    app_files = [f_sys, f_main]
    if os.path.exists(main4_path):
        with open(main4_path, "rb") as f:
            raw4 = f.read()
        main4_load_addr = raw4[0] | (raw4[1] << 8)
        f_main4 = write_file(disk, alloc, "MAIN4.BIN", 0x06,
                             main4_load_addr, raw4[4:])
        app_files.append(f_main4)
        print(f"  MAIN4.BIN {len(raw4) - 4}B (load=${main4_load_addr:X}, "
              f"key={f_main4['key_block']}, {f_main4['total_blocks']} blocks)")

    data_files = add_data_file(disk, alloc, "PCM", 0x06,
                               PCM_START_BLOCK, pcm_blocks)
    if art_blocks:
        data_files += add_data_file(disk, alloc, "ART", 0x06,
                                    ART_START_BLOCK, art_blocks)
    for e in data_files:
        print(f"  {e['name']:<8}  {e['eof']}B ({e['total_blocks']} blocks, "
              f"key={e['key_block']})")

    # ---- Rewrite the root directory (block 2) ----
    vol = memoryview(disk)[2 * BLOCK:3 * BLOCK]
    keep = []
    for i in range(1, 13):
        off = 4 + i * 39
        if vol[off] == 0:
            continue
        name_len = vol[off] & 0x0F
        name = bytes(vol[off + 1:off + 1 + name_len]).decode("ascii", "replace")
        if name in KEEP_FILES:
            keep.append(bytes(vol[off:off + 39]))
    for i in range(1, 13):
        for j in range(4 + i * 39, 4 + (i + 1) * 39):
            vol[j] = 0

    idx = 1
    for raw in keep:
        vol[4 + idx * 39:4 + idx * 39 + 39] = raw
        idx += 1
    for e in app_files + data_files:
        write_dir_entry(vol, 4 + idx * 39, e)
        idx += 1

    entry_count = len(keep) + len(app_files) + len(data_files)
    vol[0x25] = entry_count & 0xFF
    vol[0x26] = (entry_count >> 8) & 0xFF

    # ---- Mark used blocks in the ProDOS volume bitmap (block 6) ----
    # Bit set = free, bit clear = used.
    def set_used(b):
        disk[6 * BLOCK + (b // 8)] &= ~(1 << (7 - (b % 8))) & 0xFF

    for b in alloc.newly_allocated:
        set_used(b)
    for b in range(PCM_START_BLOCK, PCM_START_BLOCK + pcm_blocks):
        set_used(b)
    for b in range(ART_START_BLOCK, ART_START_BLOCK + art_blocks):
        set_used(b)

    os.makedirs(args.out_dir, exist_ok=True)
    out_path = os.path.join(args.out_dir, args.out_name)
    with open(out_path, "wb") as f:
        f.write(disk)
    print(f"\n  Built {out_path} ({len(disk)} bytes = {len(disk) // 1024} KB)")
    print(f"  Root files: {len(keep)} system + "
          + " + ".join(f["name"] for f in app_files))
    return 0


if __name__ == "__main__":
    sys.exit(main())
